storage/core/src/server/ServerStorage.ts

```diff
@@ -55,7 +55,34 @@ export abstract class ServerStorage
 
   public abstract createBaseDirectory(directory: BaseDirectory): Promise<void>;
 
-  public abstract listDirectories(): Promise<BaseDirectory[]>;
+  public async listDirectories(): Promise<BaseDirectory[]> {
+    let allDirectories: BaseDirectory[] = [];
+    const maxPageSize = 1000;
+    let continuationToken: string | undefined = undefined;
+    while (continuationToken == undefined && continuationToken != "") {
+      const res: [BaseDirectory[], string] = await this.listDirectoriesAsync(
+        maxPageSize,
+        continuationToken
+      );
+      allDirectories = [...allDirectories, ...res[0]];
+      continuationToken = res[1];
+    }
+    return allDirectories;
+  }
+
+  /**
+   * Get list of directories
+   * @param maxPageSize Max number of directories returned in the page 1000
+   * by default
+   * @param continuationToken Token to continuate current pagination
+   * (not needed in first call)
+   * @returns {Promise<[BaseDirectory[], string]>} Tuple of List of directories
+   * and continuation token, token is empty if all object were listed
+   */
+  public abstract listDirectoriesAsync(
+    maxPageSize: number,
+    continuationToken?: string
```

Comment:
We shouldn't expose implementation details (e.g., that we are using continuation token or how it looks like). There's an example of how this is implemented in [imodels-clients](https://github.com/iTwin/imodels-clients/blob/63625dcb5425b1c8076c9d947b0742b72ff32745/clients/imodels-client-management/src/base/internal/iterators/EntityPageListIterator.ts) with AsyncIterableIterator.